src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,273 @@
+    let parsed_vcek = X509Certificate::from_der(vcek_der)?.1.tbs_certificate;
+
+    // verify vcek fields
+    // chip id
+    if get_oid::<64>(&parsed_vcek, HW_ID_OID)? != evidence.attestation_report.chip_id {
+        return Err(anyhow!("Chip ID mismatch"));
+    }
+
+    // tcb version
+    // these integer extensions are 3 bytes with the last byte as the data
+    if get_oid_int(&parsed_vcek, UCODE_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.microcode
+    {
+        return Err(anyhow!("Microcode verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, SNP_SPL_OID)? != evidence.attestation_report.reported_tcb.snp {
+        return Err(anyhow!("SNP verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, TEE_SPL_OID)? != evidence.attestation_report.reported_tcb.tee {
+        return Err(anyhow!("TEE verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, LOADER_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.boot_loader
+    {
+        return Err(anyhow!("Boot loader verion mismatch"));
+    }
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29f];
+
+    sig.verify(data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn load_milan_cert_chain() -> Result<(x509::X509, x509::X509)> {
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ask_ark.pem"))?;
+
+    // ask, ark
+    Ok((certs[0].clone(), certs[1].clone()))
+}
+
+fn verify_cert_chain(cert_chain: &[CertTableEntry]) -> Result<x509::X509> {
+    let (ask, ark) = load_milan_cert_chain()?;
+
+    let raw_vcek = cert_chain
+        .iter()
+        .find(|c| c.cert_type == SnpCertType::VCEK)
+        .ok_or_else(|| anyhow!("VCEK not found."))?;
+    let vcek = x509::X509::from_der(raw_vcek.data()).map_err(|_| anyhow!("Failed to load VCEK"))?;
+
+    // ARK -> ARK
+    ark.verify(&(ark.public_key().unwrap() as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ARK Signature"))
+        .unwrap();
+
+    // ARK -> ASK
+    ask.verify(&(ark.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ASK Signature"))?;
+
+    // ASK -> VCEK
+    vcek.verify(&(ask.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid VCEK Signature"))?;
+
+    Ok(vcek)
+}
+
+fn calculate_expected_report_data(nonce: &String, tee_pubkey: &String) -> [u8; 64] {
+    let mut hasher = Sha384::new();
+
+    hasher.update(nonce.as_bytes());
+    hasher.update(tee_pubkey.as_bytes());
+
+    let mut hash = hasher.finalize().to_vec();
+    hash.extend([0u8; 16]);
+
+    hash.as_slice().try_into().unwrap()
+}
+
+fn parse_tee_evidence(report: &AttestationReport) -> Result<TeeEvidenceParsedClaim> {
+    let claims_map = json!({
+        // policy fields
+        "policy_abi_major": report.policy.abi_major(),
```

Comment:
I think those need to be string => string maps (see [here](https://github.com/confidential-containers/attestation-service/blob/main/src/lib.rs#L131)). It's maybe a bit unfortunate that the parsed claim is a untyped json value, since we deserialize the string in the same crate

Reply:
Hm I added in some formats. Not sure this is really optimal, but I will double check the values when we test the evidence parsing. Haven't quite gotten to that point yet.